crucible-llvm-syntax/src/Lang/Crucible/LLVM/Syntax.hs

```diff
@@ -0,0 +1,182 @@
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE ImplicitParams #-}
+{-# LANGUAGE ImportQualifiedPost #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE PatternSynonyms #-}
+
+module Lang.Crucible.LLVM.Syntax (llvmParserHooks) where
+
+import Control.Applicative (empty)
+import Control.Monad (unless)
+import Control.Monad.IO.Class (MonadIO(..))
+import Control.Monad.State.Strict (MonadState(..))
+import Control.Monad.Writer.Strict (MonadWriter(..))
+import Data.Functor ((<&>))
+
+import Prettyprinter (pretty)
+
+import Data.Parameterized.Context qualified as Ctx
+import Data.Parameterized.Some (Some(..))
+
+import What4.ProgramLoc (Posd(..))
+
+import Lang.Crucible.CFG.Expr qualified as Expr
+import Lang.Crucible.CFG.Reg (Atom, GlobalVar, Stmt)
+import Lang.Crucible.CFG.Reg qualified as Reg
+import Lang.Crucible.CFG.Extension (IsSyntaxExtension)
+import Lang.Crucible.Types (TypeRepr(..))
+
+import Lang.Crucible.LLVM.DataLayout (Alignment)
+import Lang.Crucible.LLVM.DataLayout qualified as DataLayout
+import Lang.Crucible.LLVM.Extension (LLVM)
+import Lang.Crucible.LLVM.Extension qualified as Ext
+import Lang.Crucible.LLVM.MemModel (Mem, pattern LLVMPointerRepr)
+import Lang.Crucible.LLVM.MemModel qualified as Mem
+import Lang.Crucible.LLVM.MemType (MemType)
+import Lang.Crucible.LLVM.MemType qualified as MemType
+import Lang.Crucible.LLVM.Translation (llvmTypeAsRepr)
+
+import Lang.Crucible.Syntax.Atoms (Atomic)
+import Lang.Crucible.Syntax.Atoms qualified as Atom
+import Lang.Crucible.Syntax.Concrete (ParserHooks(..), SyntaxState)
+import Lang.Crucible.Syntax.Concrete qualified as Parse
+import Lang.Crucible.Syntax.ExprParse (MonadSyntax)
+import Lang.Crucible.Syntax.ExprParse qualified as Parse
+
+unary :: MonadSyntax Atomic m => m b -> m a -> m a
+unary p0 p = Parse.followedBy p0 (Parse.commit *> Parse.cons p Parse.emptyList) <&> fst
```

Comment:
This is copy-pasted from `Lang.Crucible.Syntax.Concrete`, right? Is it worth exporting this function somewhere from `crucible-syntax`?

Reply:
It's close - that function takes a `Keyword` as its first argument, and `Keyword` is not extensible based on the language extension in use. We may want to revisit that in the future, parsing things as `AtomName`s and then matching on particular ones feels like a hack...

Reply:
Documented